Karl/claire.h

```diff
@@ -0,0 +1,174 @@
+
+  static std::string GenerateRandomCodename() {
+    std::string codename;
+    for (int i = 0; i < 6; ++i) {
+      codename += static_cast<char>(current::random::CSRandomInt('A', 'Z'));
+    }
+    return codename;
+  }
+
+  GenericClaire(Locator karl, const std::string& service, uint16_t port)
+      : register_called_(false),
+        registered_(false),
+        karl_(karl),
+        service_(service),
+        codename_(GenerateRandomCodename()),
+        port_(port),
+        us_start_(current::time::Now()),
+        http_scope_(HTTP(port).Register("/.current",
+                                        [this](Request r) {
+                                          const auto& qs = r.url.query;
+                                          const bool all = qs.has("all") || qs.has("a");
+                                          const bool build = qs.has("build") || qs.has("b");
+                                          if (!all && build) {
+                                            r(build::Info());
+                                          } else {
+                                            if (!registered_) {
+                                              ClaireStatusBase response;
+                                              FillBase(response, all);
+                                              r(response);
+                                            } else {
+                                              status_t response;
+                                              FillBase(response, all);
+                                              if (status_generator_) {
+                                                std::lock_guard<std::mutex> lock(mutex_);
+                                                response.runtime = status_generator_();
+                                              }
+                                              r(response);
+                                            }
+                                          }
+                                        })) {
+    // TODO(dkorolev) + TODO(mzhurovich): Should Claire at least check whether Karl is available at
+    // construction?
+  }
+
+  ~GenericClaire() {
+    if (keepalive_thread_.joinable()) {
+      keepalive_thread_.join();
+    }
+  }
+
+  void Register(status_generator_t status_filler = nullptr, bool strict = false) {
+    // Register this Claire with Karl and spawn the thread to send regular keepalives.
+    // If `strict` is true, throw if Karl can be not be reached.
+    // If `strict` is false, just start the keepalives thread.
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (!register_called_) {
+      register_called_ = true;
+      status_generator_ = status_filler;
+      // Only register once.
+      if (strict) {
+        // During this call, Karl would crawl the endpoint of this service, and, if everything is successful,
+        // register this service as the running and browsable one.
```

Comment:
I was thinking about slightly different logic in Claire:
 * Do nothing in ctor (speaking of your comment above).
 * On `Register()` send `strict` value to Karl to control its registration behavior. I.e. regardless of `strict` value we always do POST on Karl's URL and wait for `200 OK`.
   * If `strict == false`, Karl just stores the information about the service and responds 200.
   * If `strict == true`, Karl tries to check the availability of the Claire, and only then sends 200 or error, if the service is not accessible from its side.

What do you think?